src/types/middleware.ts

```diff
@@ -2,8 +2,8 @@ import InstantSearch from '../lib/InstantSearch';
 import { UiState } from './ui-state';
 import { AtLeastOne } from './utils';
 
-export type MiddlewareDefinition = {
-  onStateChange(options: { uiState: UiState }): void;
+export type MiddlewareDefinition<TUiState extends UiState = UiState> = {
+  onStateChange(options: { uiState: TUiState }): void;
```

Comment:
Isn't it supposed to be this here too?

```suggestion
export type MiddlewareDefinition<TUiState = Record<string, unknown>> = {
  onStateChange(options: { uiState: UiState & TUiState }): void;
```

Reply:
I was thinking every place except the root only needs to deal with UiState and likely no extensions (not really user code), so all except the entry point expects the consumer to make the mix

Reply:
Right, that looks fair like that!

Reply:
Right, that looks fair like that!